File: problems/7/7.c
// gcc 7.c -std=c11 -lm
//
// By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see
// that the 6th prime is 13.
//
// What is the 10 001st prime number?


#include <stdio.h>
#include <time.h>
#include <math.h>

#define TRUE 1
#define FALSE 0


int solve() {
    int prime_count = 2, number = 2, is_prime = FALSE;

    while(1) {
        // going to j <= number will make outer while endless, since is_prime
        // will always end up as FALSE. If you don't want to use sqrt(number)
        // go with < instead of <= ;)
        for(int j = 2; j <= sqrt(number); j++) {
            if(number % j == 0) {
                is_prime = FALSE;
                break;
            } else {
                is_prime = TRUE;
            }
        }

        if(is_prime) {
            prime_count++;
            if(prime_count == 10001) {
                return number;
            }
        }

        number += 1;
    }
}


int main() {
    clock_t begin = clock();
    int result = solve();
    clock_t end = clock();
    double time = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("%d [in %f seconds]", result, time);
}

File: problems/07/7.py
# coding: utf-8

"""
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that
the 6th prime is 13.

What is the 10 001st prime number?
"""


import itertools
import math
import time


def oneliner():
    # Fun fact, this 'oneliner' with names for lambdas is
    # half-an-order-of-magnitude slower than the actual oneliner that does
    # pretty much the same thing but in actual one line. shrug.jpg
    #
    # is_prime = lambda n: all(
    #     [n % j != 0
    #      for j in range(2, int(math.sqrt(n))+1)]
    # )
    # infinite_primes = (
    #     p for p in itertools.count()
    #     if is_prime(p) and p >= 2
    # )
    # return list(
    #     itertools.takewhile(
    #         lambda x: x[0] <= 10001, enumerate(infinite_primes, 1)
    #     )
    # )[-1]

    return list(
        itertools.takewhile(
            lambda x: x[0] <= 10001, enumerate(
                (p for p in itertools.count()
                if (all(p % j != 0
                       for j in range(2, int(math.sqrt(p))+1))
                    and p >= 2)),
                1  # start enumerating from 1.
            )
        )
    )[-1]


def solve():
    prime_counter, number = 0, 2
    while True:

        for j in range(2, int(math.sqrt(number))+1):
            if number % j == 0:
                break
        else:
            prime_counter += 1
            if prime_counter == 10001:
                return prime_counter, number

        number += 1


if __name__ == "__main__":
    def timeit(function):
        t1 = time.time()
        output = function()
        t2 = time.time()
        return output, t2-t1

    print timeit(solve)
    print timeit(oneliner)
